agent/agent.go

```diff
@@ -489,3 +477,50 @@ func (a *Agent) Stop() {
 func (a *Agent) addHandler(msgType string, handlerFunc handler.MessageHandlerFunc) {
 	a.handler.AddHandler(msgType, handlerFunc)
 }
+
+// StartStatsd starts up a StatsD listener on the agent, logs an error for any
+// failures.
+func (a *Agent) StartStatsd() {
```

Comment:
So I think this function is a bit more complex than it needs to be.

Consider:
```
var runStatsd func() error
if runtime.GOOS == "windows" {
    runStatsd = func() error {
        // windows specific stuff
    }
} else {
    runStatsd = func() error {
        // linux specific stuff
    }
}
if err := runStatsd(); err != nil {
    logger.WithError(err).Errorf("error running statsd on %s", addr)
}
```

There's also no need to handle errors with channels here, because the channels are not connecting different goroutines; the chan send and recv are in the same goroutine, so you can just use regular variable assignment and avoid the select.

Finally, keep in mind than an error can occur after a span of time has passed, it might not be from starting the server but another type of error. (Just for the sake of the error message being truthy)

Reply:
@echlebek Yes, I did try to simplify this, but the caveat here is that `a.statsdServer.Run(a.context)` returns an error and must be run in a goroutine. From the research I did (ex. https://www.atatus.com/blog/goroutines-error-handling/) it seemed as though retrieving an error from a goroutine using channels was the safest way.

> Whether can we directly return values from goroutines similar to a normal function? This is a big no-no in the world of concurrency.

But if this is just referring to assigning variables between _different_ goroutines, then I can definitely go ahead and simplify this, since the two goroutines are _not_ connected in this case!

Reply:
Yeah, you definitely need to start up a new goroutine to handle this stuff, just don't need to use select mechanisms and channels here. :)

Reply:
@echlebek it is done. 
![giphy](https://user-images.githubusercontent.com/10634078/39884691-4aeda714-5450-11e8-952d-7c84fecae220.gif)